python/dgl/distributed/nn/pytorch/sparse_emb.py

```diff
@@ -0,0 +1,194 @@
+    optimizers in the future.
+
+    Distributed embeddings are sharded and stored in a cluster of machines in the same way as
+    py:meth:`dgl.distributed.DistTensor`, except that distributed embeddings are trainable.
+    Because distributed embeddings are sharded
+    in the same way as nodes and edges of a distributed graph, it is usually much more
+    efficient to access than the sparse embeddings provided by the deep learning frameworks.
+
+    Parameters
+    ----------
+    num_embeddings : int
+        The number of embeddings. Currently, the number of embeddings has to be the same as
+        the number of nodes or the number of edges.
+    embedding_dim : int
+        The dimension size of embeddings.
+    name : str, optional
+        The name of the embeddings. The name can uniquely identify embeddings in a system
+        so that another NodeEmbedding object can referent to the same embeddings.
+    init_func : callable, optional
+        The function to create the initial data. If the init function is not provided,
+        the values of the embeddings are initialized to zero.
+    part_policy : PartitionPolicy, optional
+        The partition policy that assigns embeddings to different machines in the cluster.
+        Currently, it only supports node partition policy or edge partition policy.
+        The system determines the right partition policy automatically.
+
+    Examples
+    --------
+    >>> def initializer(shape, dtype):
+            arr = th.zeros(shape, dtype=dtype)
+            arr.uniform_(-1, 1)
+            return arr
+    >>> emb = dgl.distributed.nn.NodeEmbedding(g.number_of_nodes(), 10, init_func=initializer)
+    >>> optimizer = dgl.distributed.optim.SparseAdagrad([emb], lr=0.001)
+    >>> for blocks in dataloader:
+    ...     feats = emb(nids)
+    ...     loss = F.sum(feats + 1, 0)
+    ...     loss.backward()
+    ...     optimizer.step()
+
+    Note
+    ----
+    When a ``NodeEmbedding``  object is used when the deep learning framework is recording
+    the forward computation, users have to invoke
+    py:meth:`~dgl.distributed.optim.SparseAdagrad.step` afterwards. Otherwise, there will be
+    some memory leak.
+    '''
+    def __init__(self, num_embeddings, embedding_dim, name=None,
+                 init_func=None, part_policy=None):
+        self._tensor = DistTensor((num_embeddings, embedding_dim), F.float32, name,
+                                  init_func=init_func, part_policy=part_policy)
+        self._trace = []
+        self._name = name
+        self._num_embeddings = num_embeddings
+        self._embedding_dim = embedding_dim
+
+        # Check whether it is multi-gpu/distributed training or not
+        if th.distributed.is_initialized():
+            self._rank = th.distributed.get_rank()
+            self._world_size = th.distributed.get_world_size()
+        else:
+            assert 'th.distributed shoud be initialized'
+        self._optm_state = None # track optimizer state
```

Comment:
it's interesting that you choose to keep track of the optimizer state in NodeEmbedding, instead of hiding it inside the optimizer. Is it necessary to keep track of it in the embedding?

Reply:
Good point. Actually, we also store the status in the optimizer. We do not need it here.